harness/determined/pytorch/samplers.py

```diff
@@ -0,0 +1,361 @@
+  whenever possible, to achieve the highest-quality shuffle.
+
+- **Repeat when training**: In Determined, you always repeat your training dataset and you
+  never repeat your validation datasets.  Determined provides a RepeatSampler and a
+  RepeatBatchSampler to wrap your sampler or batch_sampler.  For your training dataset, make sure
+  that you always repeat AFTER you shuffle, otherwise your shuffle will hang.
+
+- **Always shard, and not before a repeat**: Use Determined's DistributedSampler or
+  DistributedBatchSampler to provide a unique shard of data to each worker based on your sampler or
+  batch_sampler.  It is best to always shard your data, and even when you are not doing distributed
+  training, because in non-distributed-training settings, the sharding is nearly zero-cost, and it
+  makes distributed training seamless if you ever want to use it in the future.
+
+  It is generally important to shard after you repeat, unless you can guarantee that each shard
+  of the dataset will have the same length.  Otherwise, differences between the epoch boundaries for
+  each worker can grow over time, especially on small datasets.  If you shard after you repeat, you
+  can change the number of workers arbitrarily without issue.
+
+- **Skip when training, and always last**: In Determined, training datasets should always be able to
+  start from an arbitrary point in the dataset.  This allows for advanced hyperparameter searches
+  and responsive preemption for training on spot instances in the cloud.  The easiest way to do
+  this, which is also very efficient, is to apply a skip to the sampler.
+
+  Determined provides a SkipBatchSampler that you can apply to your batch_sampler for this purpose.
+  There is also a SkipSampler that you can apply to your sampler, but you should prefer to skip on
+  batches unless you are confident that your dataset always yields identical size batches, where the
+  number of records to skip can be reliably calculatd from the number of batches already trained.
+
+  Always skip AFTER your repeat, so that the skip only happens once, and not on every dataset.
+
+  Always skip AFTER your shuffle, to preserve the reproducibility of the shuffle.
+
+Here is some example code that follows each of these rules that you can use as a starting point if
+you find that the built-in context.DataLoader() does not support your use case.
+
+  .. code::python
+
+    def make_batch_sampler(
+        sampler_or_dataset,
+        shuffle,
+        seed,
+        repeat,
+        num_workers,
+        rank,
+        batch_size,
+        skip,
+    ):
+        if isinstance(sampler_or_dataset, torch.utils.data.Sampler):
+            sampler = sampler_or_dataset
+        else:
+            # Create a SequentialSampler if we started with a Dataset.
+            sampler = torch.utils.data.SequentialSampler(sampler_or_dataset)
+
+        # Shuffle first.
+        if shuffle:
```

Comment:
its a bit confusing in this example with all the else ifs. Didn't we say above that you really should use all these steps?

Reply:
good point, I'll clear this code example up with a `mode` corresponding to `"training"` or `"validation"`